### ISSUE TITLE
Weird Response when set both "Template and Redirect" on same route as return value type

### ISSUE THREAD

**Issue by 4n1qz5skwv** [litestar-org/litestar#1652]
### Discussed in https://github.com/orgs/litestar-org/discussions/1646

<div type='discussions-op-text'>

<sup>Originally posted by **4n1qz5skwv** May 10, 2023</sup>
```python
from pathlib import Path

from litestar import Litestar, Request, get, HttpMethod, route
from litestar.contrib.jinja import JinjaTemplateEngine
from litestar.response_containers import Template, Redirect
from litestar.template.config import TemplateConfig


@get("/home")
async def aaa() -> str:
    return 'home!'


@route(path="/some-path", http_method=[HttpMethod.GET, HttpMethod.POST])
async def bbb(request: Request) -> Template | Redirect:
    if request.method == "POST":
        # may be login and move to home
        return Redirect(path='/home', status_code=302)

    # show template on opening page
    return Template(name="abc.html")


app = Litestar(route_handlers=[aaa, bbb],
               template_config=TemplateConfig(directory=Path(__file__).parent, engine=JinjaTemplateEngine))

if __name__ == '__main__':
    import uvicorn

    uvicorn.run('0:app')
```

**curl**
```
Microsoft Windows [Version 10.0.19044.2846]
(c) Microsoft Corporation. All rights reserved.

C:\Users\User-PC>curl http://127.0.0.1:8000/some-path -X GET
{"name":"abc.html","context":{},"background":null,"headers":{},"cookies":[],"media_type":null,"encoding":"utf-8"}
C:\Users\User-PC>curl http://127.0.0.1:8000/some-path -X POST
{"path":"/home","background":null,"headers":{},"cookies":[],"media_type":null,"encoding":"utf-8","status_code":302}
C:\Users\User-PC>
```

**logs**
```
INFO:     Started server process [14912]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:52708 - "GET /some-path HTTP/1.1" 201 Created
INFO:     127.0.0.1:52709 - "POST /some-path HTTP/1.1" 201 Created
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [14912]

```

</div>

**Comment by peterschutt** [comment]
This case is fairly straightforward if we consider that a union is a subtype of `X` if all inner types of the union are also subtypes of `X`, which on the surface seems OK to me.

Specific to this case, can we say that `Template | Redirect` is a subtype of `Response` because both `Template` and `Redirect` _are_ subtypes of `Response`.

This is relevant because:

 https://github.com/litestar-org/litestar/blob/f27b33c2142d64c3ec660b2f5f194803100550ae/litestar/handlers/http_handlers/base.py#L424-L427

At the moment, `return_type.is_subclass_of(Response)` is returning `False` - however, if that is allowed to return `True` than this "just works" because both `Template` and `Redirect` are serviced by the same response handler function.

It was a deliberate choice at the time to have any union type return `False` from `ParsedType.is_subclass_of()` with the intention of waiting for some real world cases to arrive - now we have one.

@litestar-org/maintainers WDT? Is `Union[SubclassOfX1, SubclassOfX2]` a subtype of `X`?

**Comment by Goldziher** [comment]
> This case is fairly straightforward if we consider that a union is a subtype of `X` if all inner types of the union are also subtypes of `X`, which on the surface seems OK to me.
> 
> Specific to this case, can we say that `Template | Redirect` is a subtype of `Response` because both `Template` and `Redirect` _are_ subtypes of `Response`.
> 
> This is relevant because:
> 
>  https://github.com/litestar-org/litestar/blob/f27b33c2142d64c3ec660b2f5f194803100550ae/litestar/handlers/http_handlers/base.py#L424-L427
> 
> At the moment, `return_type.is_subclass_of(Response)` is returning `False` - however, if that is allowed to return `True` than this "just works" because both `Template` and `Redirect` are serviced by the same response handler function.
> 
> It was a deliberate choice at the time to have any union type return `False` from `ParsedType.is_subclass_of()` with the intention of waiting for some real world cases to arrive - now we have one.
> 
> @litestar-org/maintainers WDT? Is `Union[SubclassOfX1, SubclassOfX2]` a subtype of `X`?

Sure

**Comment by provinzkraut** [comment]
I'm a bit lost here. Is this not a duplicate of #1371, which was already fixed?

**Comment by peterschutt** [comment]
Perhaps that PR occurred before we started using `ParsedType.is_subtype_of()` when inspecting the return types?

**Comment by peterschutt** [comment]
The specific issue here is that response type `Template | Redirect` doesn't evaluate to `True` here: https://github.com/litestar-org/litestar/blob/f27b33c2142d64c3ec660b2f5f194803100550ae/litestar/handlers/http_handlers/base.py#L424

So the response from the handler is processed with the usual data handler.

**Comment by peterschutt** [comment]
We still have a case where something like this won't work:

```py
@get()
def handler() -> Response | str:
    return random.choice(Response("string"), "string")
```

I think we'd either need a special union response handler for that case, or we disallow it.

**Comment by provinzkraut** [comment]
> Perhaps that PR occurred before we started using `ParsedType.is_subtype_of()` when inspecting the return types?

Probably. I didn't add a specific test case for the issue back then, since the fix was unrelated to the type annotations and basically just to allow using a completely different pattern, so it seems like this is simply a regression in the return annotation handling.

**Comment by peterschutt** [comment]
Happy for #1690 then?

**Comment by provinzkraut** [comment]
It seems to solve this issue, so let's go with that for now!

As your example above demonstrates though, there are still other cases that aren't handled correctly. But that probably should be addressed elsewhere and is a more complicated issue.
